lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -272,225 +317,422 @@ static Operation *checkSubModuleOp(mlir::ModuleOp parentModule,
+    outputOp->moveAfter(&parent.getBodyBlock()->back());
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+template <typename T>
+class ExtModuleConversionPattern : public OpConversionPattern<T> {
+public:
+  ExtModuleConversionPattern(ESITypeConverter &typeConverter,
+                             MLIRContext *context, OpBuilder &submoduleBuilder,
+                             HandshakeLoweringState &ls)
+      : OpConversionPattern<T>::OpConversionPattern(typeConverter, context),
+        submoduleBuilder(submoduleBuilder), ls(ls) {}
+  using OpAdaptor = typename T::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(T op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    hw::HWModuleLike implModule = checkSubModuleOp(ls.parentModule, op);
+    if (!implModule) {
+      auto portInfo = ModulePortInfo(getPortInfoForOp(rewriter, op));
+      implModule = submoduleBuilder.create<hw::HWModuleExternOp>(
+          op.getLoc(), submoduleBuilder.getStringAttr(getSubModuleName(op)),
+          portInfo);
     }
-    rewriter.eraseOp(funcOp);
+
+    llvm::SmallVector<Value> operands = adaptor.getOperands();
+    addSequentialIOOperandsIfNeeded(op, operands);
+    rewriter.replaceOpWithNewOp<hw::InstanceOp>(
+        op, implModule, rewriter.getStringAttr(ls.nameUniquer(op)), operands);
     return success();
   }
 
 private:
-  /// Maintain a map from module names to the # of times the module has been
-  /// instantiated inside this module. This is used to generate unique names for
-  /// each instance.
-  mutable std::map<std::string, unsigned> instanceNameCntr;
+  OpBuilder &submoduleBuilder;
+  HandshakeLoweringState &ls;
 };
 
+class FuncOpConversionPattern : public OpConversionPattern<handshake::FuncOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(handshake::FuncOp op, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    ModulePortInfo ports = getPortInfoForOp(rewriter, op, op.getArgumentTypes(),
+                                            op.getResultTypes());
+    auto hwModule = rewriter.create<hw::HWModuleOp>(
+        op.getLoc(), rewriter.getStringAttr(op.getName()), ports);
+    auto args = hwModule.getArguments().drop_back().drop_back();
+    rewriter.mergeBlockBefore(&op.getBody().front(),
+                              hwModule.getBodyBlock()->getTerminator(), args);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// HW Top-module Related Functions
+//===----------------------------------------------------------------------===//
+
+static bool isMemrefType(Type t) { return t.isa<mlir::MemRefType>(); }
+static LogicalResult verifyHandshakeFuncOp(handshake::FuncOp &funcOp) {
+  // @TODO: memory I/O is not yet supported. Figure out how to support memory
```

Comment:
It's coming! There's nothing in tree, but for the HotChips demo we prototyped something purely in PyCDE, where the "memory" responses were sent by Python during cosimulation.